src/Compilers/Core/Portable/InternalUtilities/SemaphoreSlimExtensions.cs

```diff
@@ -16,8 +16,13 @@ public static SemaphoreDisposer DisposableWait(this SemaphoreSlim semaphore, Can
 
         public async static Task<SemaphoreDisposer> DisposableWaitAsync(this SemaphoreSlim semaphore, CancellationToken cancellationToken = default)
```

Comment:
should this be `ValueTask<SemaphoreDisposer>`.  Same question with a lot of your other changes?

Reply:
Could be. But given `await SemaphoreSlim.WaitAsync()` is not going to finish synchronously, this won't finish synchronously, so we still end up creating a Task.

Reply:
sure, in the async case, we'd allocate.  But my gut makes me think in most cases we wouldn't be contentious.  but maybe that should be examined in another PR :)

Reply:
Yeah, we'll get some benchmarks for this to see how it performs.